components/org.wso2.carbon.identity.outbound.provisioning.connector.office365/src/main/java/org/wso2/carbon/identity/outbound/provisioning/connector/office365/Office365ProvisioningConnector.java

```diff
@@ -0,0 +1,422 @@
+        if (ruleAttributeClaim.isEmpty() && !ruleAttributeName.isEmpty()) {
+            ruleAttributeClaim = Office365ConnectorConstants.WSO2_ROLE_CLAIM;
+        }
+        String displayName = requiredAttributes.get(displayNameClaim);
+        String mailNickName = requiredAttributes.get(mailNickNameClaim);
+        String immutableId = requiredAttributes.get(immutableIdClaim);
+        String upn = requiredAttributes.get(upnClaim);
+        String ruleAttributeValue = requiredAttributes.get(ruleAttributeClaim);
+
+        if (displayName == null || mailNickName == null || immutableId == null || upn == null) {
+            throw new NullPointerException("One or more of the mandatory user attributes: display name, mail " +
+                    "nickname, immutable id, user principal name do not have a value.");
+        } else {
+            // Create a json object corresponding to the attributes of the user in the request.
+            JSONObject passwordProfile = new JSONObject();
+            passwordProfile.put(Office365ConnectorConstants.FORCE_CHANGE_PASSWORD, false);
+            passwordProfile.put(Office365ConnectorConstants.PASSWORD, RandomStringUtils.randomAlphanumeric(12));
+
+            JSONObject user = new JSONObject();
+            user.put(Office365ConnectorConstants.ACCOUNT_ENABLED, true);
+            user.put(Office365ConnectorConstants.OFFICE365_DISPLAY_NAME, displayName);
+            user.put(Office365ConnectorConstants.OFFICE365_EMAIL_NICKNAME, mailNickName);
+            user.put(Office365ConnectorConstants.OFFICE365_UPN, getDomainSpecificUpn(upn));
+            user.put(Office365ConnectorConstants.OFFICE365_IMMUTABLE_ID, immutableId);
+            user.put(Office365ConnectorConstants.PASSWORD_PROFILE, passwordProfile);
+            if (!ruleAttributeName.isEmpty()) {
+                user.put(ruleAttributeName, ruleAttributeValue);
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("A user object is created. " + user.toString());
+            }
+            return user;
+        }
+    }
+
+    private void setAuthorizationHeader(HttpRequestBase httpMethod) throws IdentityProvisioningException {
+
+        String accessToken = getAccessToken();
+
+        if (!accessToken.isEmpty()) {
+            httpMethod.addHeader(Office365ConnectorConstants.AUTHORIZATION_HEADER_NAME,
+                    Office365ConnectorConstants.AUTHORIZATION_HEADER_BEARER + " " + accessToken);
+
+            if (log.isDebugEnabled() && IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.ACCESS_TOKEN)) {
+                log.debug("Setting authorization header for method: " + httpMethod.getMethod() + " as follows,");
+                Header authorizationHeader = httpMethod
+                        .getLastHeader(Office365ConnectorConstants.AUTHORIZATION_HEADER_NAME);
+                log.debug(authorizationHeader.getName() + ": " + authorizationHeader.getValue());
```

Comment:
Without logging the token, let's log the hash of it.

Reply:
The token is logged only if the token loggable is enabled. So shall we keep it like that?